streams/dlna_meta.py

```diff
@@ -1,68 +1,212 @@
+
+          if transport_state == "TRANSITIONING": # metadata is empty while transitioning so just skip this state
+            continue
+
+          if transport_state != "STOPPED": # reset the stop counter if the state is not stopped
+            stop_counter = 0
+
+        except Exception as e:
+          print(f"Error: could not get transport state: {e}")
+
+        # try to get song-info from the service and parse it
+        try:
+          meta_xml = ET.fromstring(service.GetMediaInfo(InstanceID=0)["CurrentURIMetaData"])
+          metadata["title"] = ""
+          metadata["artist"] = ""
+          metadata["album"] = ""
+          for i in meta_xml.iter():
+            if i.tag == "{http://purl.org/dc/elements/1.1/}title":
+              metadata["title"] = i.text
+            elif i.tag == "{urn:schemas-upnp-org:metadata-1-0/upnp/}artist":
+              metadata["artist"] = i.text
+            elif i.tag == "{urn:schemas-upnp-org:metadata-1-0/upnp/}album":
+              metadata["album"] = i.text
+            elif i.tag == "{urn:schemas-upnp-org:metadata-1-0/upnp/}albumArtURI":
+              fname = i.text.split('/')[-1]
+              if debug:
+                print(f"Downloading album art {i.text} to {album_art_dir}/{fname}")
+
+              # move to the album art directory and download the file (if it changed), if it fails delete the file
+              os.system(f"cd {album_art_dir} && wget -q -N {i.text} || rm -f {album_art_dir}/{fname}")
+
+              # if the file exists, set the metadata to the file name
+              if os.path.exists(f"{album_art_dir}/{fname}"):
+                if fname != last_file: # if it's a new file delete the old one
+                  if last_file: # handle the case where this is the first file downloaded
+                    os.remove(f"{album_art_dir}/{last_file}")
+                  last_file = fname
+
+                metadata["album_art"] = fname
+              else:
+                metadata["album_art"] = ""
+        except Exception as e:
+          print(f"Error: could not get song-info: {e}")
 
-# Check if the folder is valid and writeable
-if not os.path.exists(args.dlna_config_dir) or not os.access(args.dlna_config_dir, os.W_OK):
-  print('Error: unable to access: {}'.format(args.dlna_config_dir))
-
-# trim trailing /
-if args.dlna_config_dir[-1] == '/':
-  args.dlna_config_dir = args.dlna_config_dir[:-1]
-
-cs_loc = '{}/currentSong'.format(args.dlna_config_dir)
-cs_conf = {
-    'state': 'playing'
-}
-ctmd = re.compile(r'dc:(.*?)>(.*?)</dc:|upnp:(.*?)>(.*?)</upnp:')
-ts = re.compile(r'TransportState: ([A-Z]*)')
-
-def read_field():
-  line = sys.stdin.readline()
-  line = line.strip('\n')
-  if line[0:4] == "INFO":
-    s1 = line.split('] ')
-    return s1[1]
+        if debug:
+          print("writing metadata", metadata)
+
+        # empty file and write metadata
+        f.truncate(0)
+        f.seek(0)
+        json.dump(metadata, f)
+        f.flush()
```

Comment:
there's not necessarily a TODO here, but file churn like this every 3s on the MMC instead of a tmpfs or similar sketches me out. I'm reading that `flush()` doesn't necessarily mean data ends up on disk (`sync`/`fsync` syscalls do that), but it still reads like this is gonna hit the MMC a ton over time. if you have any quick and clever ideas on where to stash this stuff, such that source configs persist but metadata for these sources does not, and without adding yet another line to fstab, I'm v game

Reply:
maybe the answer is literally to include a reference to the metadata file used in the persistent source config, stash it in /tmp, and handle the case where it disappears?

Reply:
I was under the impression these were in a tempfs

Reply:
this is actually in a tmpfs already, @klay2000 pointed out. nevermind :upside_down_face: